### ISSUE TITLE
Mf helper issue

### ISSUE THREAD

**Issue by oldfcker** [kolton/d2bot-with-kolbot#1373]

**Comment by DarkHorseDre** [comment]
Yeah, problem is tombs all have same name bud different ID so get bot to give id in command.

in `attack.js`, change:
`say("clearlevel " + getArea().name);`

to:
`say("clearlevel " + me.area);` //dhd enable unique name for tombs

**Comment by oldfcker** [comment]
> Yeah, problem is tombs all have same name but different ID so get bot to give id in command.
> in attack.js, change:
> say("clearlevel " + getArea().name);
> to:
> say("clearlevel " + me.area); //dhd enable unique name for tombs

this worked wonderfully thank you! now my issue is that my leachers will run out of tps and then leave game what can I do to have them do town chores inbetween acts?

  Error in MFHelper (pather.js #1032) makePortal: No scrolls. 

this is the error I get and then the champ in question will leave the game.
I am running 7 in one game with mf helper as lead setup it is working great other then this issue and 
 Cannot process request because the process (14504) has exited.
this error will cause the champ in question to crash and not return until stoped and restarted
please lmk what I can do to fix these two issues

**Comment by DarkHorseDre** [comment]
Multiple issues there, and more to come!

FIRSTLY: ensure you're running the latest MF helper as that has been updated heavily.  I haven't tried it yet but will soon.

tps are easy - add a tome and add scrolls to pickit - they should return to town and do chores to restock.  They will do it at start of every run and if you are doing all tombs you shouldn't use 20 scrolls.

the other stuff you mentioned I don't understand but wont ask you to explain until you confirm you are using the latest MFhelper.js

**Comment by oldfcker** [comment]
> Multiple issues there, and more to come!
> FIRSTLY: ensure you're running the latest MF helper as that has been updated heavily.  I haven't tried it yet but will soon.
> tps are easy - add a tome and add scrolls to pickit - they should return to town and do chores to restock.  They will do it at start of every run and if you are doing all tombs you shouldn't use 20 scrolls.
> the other stuff you mentioned I don't understand but wont ask you to explain until you confirm you are using the latest MFhelper.js

yes I have the most up to date everything. just dled it one day back as stated in my first post. I am running every spot that has mobs a1-a5 and am running out of tps on some near the end of the tombs and then the champ in question exits the game. on ng it refills tps and is fine but then a dif toon has that issue
and the error 1404 is random as heck

**Comment by DarkHorseDre** [comment]
so you have to tomb?
I do have it in pickit to help it stay full whilst running. maybe helpers don't execute town do chores when in mfhelper? I'd have to check tomorrow. 

for now add tps to pickit limited to 20 and if I'm right that helpers don't execute town do chores during a run then you're gonna have to force them to goto town at the start of long runs with lots of tp to town like in tombs.

**Comment by oldfcker** [comment]
> so you have to tomb?
> I do have it in pickit to help it stay full whilst running. maybe helpers don't execute town do chores when in mfhelper? I'd have to check tomorrow.
> for now add tps to pickit limited to 20 and if I'm right that helpers don't execute town do chores during a run then you're gonna have to force them to goto town at the start of long runs with lots of tp to town like in tombs.

yes all 7 have books with tp and id keys ect
ill try the book of tp in picket and see what happens
no they do not do town chores during runs
except when lead goes to start chaos then they do. but chaos and baal are at the end of the run and they run out in tombs almost as if they didn't total refill at the start of the run

**Comment by oldfcker** [comment]
> so you have to tomb?
> I do have it in pickit to help it stay full whilst running. maybe helpers don't execute town do chores when in mfhelper? I'd have to check tomorrow.
> for now add tps to pickit limited to 20 and if I'm right that helpers don't execute town do chores during a run then you're gonna have to force them to goto town at the start of long runs with lots of tp to town like in tombs.

tp in picket seems to have fixed it for now not getting that error at all anymore and the are filling tomes during runs via picket. other then random crashes everything is running smooth. 
they don't actually ftj but it prints it to manager.
and its never just one and wait for a while lol its like one then that one gets back to game then down goes another lol but they get right back in there so all good!

Thank you for the help!!

16:44:52 (sorc) Window has unexpectedly exited... starting profile
16:44:56 (sorc) Crash Info: Script: no entry Area: Kurast Docktown
16:45:17 (sorc) Failed to join game
16:48:38 (din2) Window has unexpectedly exited... starting profile
16:48:41 (din2) Crash Info: Script: no entry Area: Harrogath
16:49:02 (din2) Failed to join game
16:55:04 (zon) Window has unexpectedly exited... starting profile
16:55:07 (zon) Crash Info: Script: no entry Area: Icy Cellar
16:55:28 (zon) Failed to join game
17:02:29 (din) Transmuting: Ring + Sol Rune + Jewel of Fervor + Perfect Ruby
17:23:44 (din) Transmuting: Flawless Ruby + Flawless Ruby + Flawless Ruby
17:28:21 (zeal) Window has unexpectedly exited... starting profile
17:28:24 (zeal) Crash Info: Script: no entry Area: Tal Rashas Tomb #3
17:28:51 (zeal) Failed to join game

**Comment by DarkHorseDre** [comment]
So as I said, the issue is that they are using more tps than they have. they may also not be filling the TOME at the start of the run? you need to confirm that as they should.

You need to monitor your bot - things like this shouldn't be in question: start the run, pause the leader, check the followers number of tps (or simply have them print to console number of scrolls) then see what happens when they go to shop.

As for a fix, just make them do town chores in mfhelper when the leader goes to act 2.
at the start of mfhelper it checks the area of the leader and then moves to that act - so you can copy that code to make an IF statement and then do what the other scripts do to force a shop:

Town.doChores();

That line forces a visit to town if not in town and also runs the command `this.fillTome(518)`.
Have that happen before a run that requires many tps and youre done.  tbh they should only need to fill once a run - I have no idea how youre using so many tps for followers - but as you've seen, actually picking ones laying on the ground helps, but thats also RNG based, better to shop them

as for crashes - ensure you have join delays (via follow.dbj) and exiting also can cause RD (8 clients exiting to channel at the same time).  Here's how I do the quit delay, thanks to @zombie: https://github.com/kolton/d2bot-with-kolbot/issues/1331

**Comment by oldfcker** [comment]
> So as I said, the issue is that they are using more tps than they have. they may also not be filling the TOME at the start of the run? you need to confirm that as they should.
> You need to monitor your bot - things like this shouldn't be in question: start the run, pause the leader, check the followers number of tps (or simply have them print to console number of scrolls) then see what happens when they go to shop.
> As for a fix, just make them do town chores in mfhelper when the leader goes to act 2.
> at the start of mfhelper it checks the area of the leader and then moves to that act - so you can copy that code to make an IF statement and then do what the other scripts do to force a shop:
> Town.doChores();
> That line forces a visit to town if not in town and also runs the command this.fillTome(518).
> Have that happen before a run that requires many tps and youre done.  tbh they should only need to fill once a run - I have no idea how youre using so many tps for followers - but as you've seen, actually picking ones laying on the ground helps, but thats also RNG based, better to shop them
> as for crashes - ensure you have join delays (via follow.dbj) and exiting also can cause RD (8 clients exiting to channel at the same time).  Here's how I do the quit delay, thanks to @zombie: #1331

they do fill at start of the run but what I feel is happening is this line 
		if (Config.PublicMode) {
			Pather.makePortal();
it causes the helpers and all leachers to enter their own tp on returning from say buying pots and to instantly put up another tp even tho they then tele off and kill in the section they are in and on finishing that area put up another tp so in say one tomb they can use a number of tps causing them to run out if not picking or buying. the purchasing between runs I will implement but as I do watch them constantly that is the only thing that I can see that would have cause this issue. as for join issues not having any or returning to channel it is just the random crashes and rejoins in the middle of runs that are causing my RDs usually it is more then one that will do it not at same time but one will then will return to the game and then couple seconds later second will and then a third and come next game rd I guess I could just buy more cd keys and have them cycle through but in order to avoid doing so my runs are over an hour long with 7 running in game. on runs were there are but maby one crash or none things run smooth as can be from run to run. I have a comp capable of running more then 7 but wanted to leave a slot open for a friend when he returns. only thing I have as an issue is the isp but that will be fixed today.

**Comment by DarkHorseDre** [comment]
wow that's a wall of text with no breaks! I read the first part, you seem to be confirming that they fill a tome and are using more than 20 tps in a run - you know how to fix that now, as above.

You should be good to go now - goodluck bro
